File: Fitness/ViewControllers/TrainingTableViewController.m
//
//  TrainingTableViewController.m
//  Fitness
//
//  Created by Mani on 23.04.15.
//  Copyright (c) 2015 Mani. All rights reserved.
//

#import <ResearchKit/ResearchKit.h>
#import "TrainingTableViewController.h"
#import "EditTrainingTableViewController.h"
#import "../Health/Health.h"
#import "../Resources/Training.h"
#import "../Resources/Exercise.h"
#import "../Resources/Storage.h"

@interface TrainingTableViewController () <EditTrainingTableViewControllerDelegate, ORKTaskViewControllerDelegate>
@property(strong) NSMutableArray *trainings;
@property NSDate *trainingStart;
@property NSInteger lastStep;
@property UILocalNotification *currentNotification;

- (void)startLocalNotificationWithInterval:(NSTimeInterval)interval;
- (void)cancelLocalNotification;

@end

@implementation TrainingTableViewController

#pragma mark Overloaded Base Methods
- (void)viewDidLoad {
    [super viewDidLoad];

    self.navigationItem.leftBarButtonItem = self.editButtonItem;
    self.tableView.allowsSelectionDuringEditing = YES;
    
    // Load the trainings
    self.trainings = [Storage trainings];
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
}

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if([segue.destinationViewController isKindOfClass:[UINavigationController class]]) {
        UIViewController *underlayingController = ((UINavigationController *)segue.destinationViewController).topViewController;
        
        if([underlayingController isKindOfClass:[EditTrainingTableViewController class]]) {
            EditTrainingTableViewController *controller = (EditTrainingTableViewController *)underlayingController;
            
            controller.delegate = self;
            controller.sender = sender;
            
            // Edit training
            if([sender isKindOfClass:[UITableViewCell class]]) {
                NSInteger index = [self.tableView indexPathForCell:(UITableViewCell *)sender].row;
                controller.training = [[self.trainings objectAtIndex:index] mutableCopy];
            }
            else {
                // Insert training
                controller.training = [[Training alloc] init];
            }
        }
    }
}

#pragma mark UITableView Data Source
- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.trainings.count;
}

#pragma mark UITableView Delegation
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"TrainingTableViewCell" forIndexPath:indexPath];
    
    Training *training = [self.trainings objectAtIndex:indexPath.row];
    cell.textLabel.text = training.name;
    
    NSInteger durationMinutes = ((NSInteger)[training exerciseDuration]) / 60;
    NSInteger durationSeconds = ((NSInteger)[training exerciseDuration]) % 60;
    
    cell.detailTextLabel.text = [NSString stringWithFormat:@"%ld Übung%@, Dauer: %ld Minute%@ %ld Sekunde%@",
                                 training.exercises.count, training.exercises.count == 1 ? @"" : @"en",
                                 durationMinutes, durationMinutes == 1 ? @"" : @"n",
                                 durationSeconds, durationSeconds == 1 ? @"" : @"n"
                                 ];
    
    return cell;
}

- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle != UITableViewCellEditingStyleDelete)
        return;
    
    [self.trainings removeObjectAtIndex:indexPath.row];
    [Storage saveTrainings:self.trainings];
    
    [tableView deleteRowsAtIndexPaths:@[indexPath] withRowAnimation: UITableViewRowAnimationFade];
}

- (void)tableView:(UITableView *)tableView moveRowAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath {
    id movee = [self.trainings objectAtIndex:fromIndexPath.row];
    [self.trainings removeObject:movee];
    [self.trainings insertObject:movee atIndex:toIndexPath.row];
    [Storage saveTrainings:self.trainings];
}


- (BOOL)tableView:(UITableView *)tableView canMoveRowAtIndexPath:(NSIndexPath *)indexPath {
    return YES;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    if(tableView.isEditing) {
        [self performSegueWithIdentifier:@"EditTrainingTableViewController" sender:[tableView cellForRowAtIndexPath:indexPath]];
    } else {
        Training *training = [self.trainings objectAtIndex:indexPath.row];
        ORKTaskViewController *controller = [[ORKTaskViewController alloc] initWithTask:[training trainingTask] taskRunUUID:nil];
        
        controller.delegate = self;
        self.trainingStart = [NSDate date];
        self.lastStep = 0;
        
        [self presentViewController:controller animated:YES completion:nil];
    }
}

#pragma mark EditTrainingTableViewController Delegation
- (void)editTrainingTableViewControllerDidCancelled:(EditTrainingTableViewController *)controller {
    [controller dismissViewControllerAnimated:YES completion:nil];
}

- (void)editTrainingTableViewControllerDidFinished:(EditTrainingTableViewController *)controller {
    if(controller.sender == nil)
        return;
    
    // The controller was opened by a table cell
    if([controller.sender isKindOfClass:[UITableViewCell class]]) {
        NSUInteger index = [self.tableView indexPathForCell:(UITableViewCell *)controller.sender].row;
        [self.trainings setObject:controller.training atIndexedSubscript:index];
                            
        [self.tableView deselectRowAtIndexPath:[self.tableView indexPathForCell:(UITableViewCell *)controller.sender] animated:YES];
    } else {
        // Store the new training and appent it at the end
        [self.trainings addObject:controller.training];
        
    }
                            
    [self.tableView reloadData];
    [controller dismissViewControllerAnimated:YES completion:nil];
    [Storage saveTrainings:self.trainings];
}

#pragma mark ORKTaskViewController Delegation
- (void)taskViewController:(ORKTaskViewController * __nonnull)taskViewController didFinishWithReason:(ORKTaskViewControllerFinishReason)reason error:(nullable NSError *)error {
    [taskViewController dismissViewControllerAnimated:YES completion:nil];
    
    // Add data to HealthKit
    if(reason == ORKTaskViewControllerFinishReasonCompleted) {
        [[Health health] addTraining:self.trainingStart end:[NSDate date] completion:nil];
        [self startLocalNotificationWithInterval:0];
    }
    
    self.lastStep = 0;
}

- (void)taskViewController:(ORKTaskViewController * __nonnull)taskViewController stepViewControllerWillAppear:(ORKStepViewController * __nonnull)stepViewController {
    ORKTaskProgress progress = [taskViewController.task progressOfCurrentStep:stepViewController.step withResult:taskViewController.result];
    if(progress.current <= self.lastStep)
        [self cancelLocalNotification];
    
    self.lastStep = progress.current;
    
    if(stepViewController.step != nil) {
        if([stepViewController.step isKindOfClass:[ORKActiveStep class]]) {
            [self startLocalNotificationWithInterval:((ORKActiveStep *)stepViewController.step).stepDuration];
        }
    }
}

#pragma mark Methods
- (void)startLocalNotificationWithInterval:(NSTimeInterval)interval {
    self.currentNotification = [UILocalNotification new];
    self.currentNotification.fireDate = [NSDate dateWithTimeIntervalSinceNow:interval];
    self.currentNotification.alertBody = @"Zeit abgelaufen!";
    self.currentNotification.soundName = UILocalNotificationDefaultSoundName;
    
    if(interval == 0)
        [[UIApplication sharedApplication] presentLocalNotificationNow:self.currentNotification];
    else
        [[UIApplication sharedApplication] scheduleLocalNotification:self.currentNotification];
}

- (void)cancelLocalNotification {
    if(self.currentNotification == nil)
        return;
    
    [[UIApplication sharedApplication] cancelLocalNotification:self.currentNotification];
    self.currentNotification = nil;
}
@end

File: Fitness/ViewControllers/TrainingTableViewController.swift
//
//  TrainingTableViewController.swift
//  Fitness
//
//  Created by Manuel Leitold on 11.09.15.
//  Copyright © 2015 - 2016 mani1337. All rights reserved.
//

import UIKit

class TrainingTableViewController: UITableViewController, InputTableViewCellDelegate {
    
    var training: Training!
    var activeTrainingController: ActiveTrainingController?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.tableView.isEditing = true
        self.tableView.allowsSelectionDuringEditing = true
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        CoreData.save()
        self.tableView.reloadData()
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        CoreData.save()
    }
    
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 3
    }

    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return [1, self.training.exercises!.count, 1][section]
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell
        
        switch indexPath.section {
        case 0:
            let newCell = tableView.dequeueReusableCell(withIdentifier: "InputCell", for: indexPath) as! InputTableViewCell
            
            newCell.delegate = self
            newCell.textField.text = self.training.name
            
            cell = newCell
            
        case 1:
            let exercise = self.training.exercises!.object(at: indexPath.row) as! Exercise
            let newCell = tableView.dequeueReusableCell(withIdentifier: "ExerciseCell", for: indexPath)
            
            newCell.textLabel?.text = ExerciseType(rawValue: exercise.exerciseType)?.localizedName()
            newCell.detailTextLabel?.text = Int(exercise.duration).asTime
                
            cell = newCell
            
        case 2:
            cell = tableView.dequeueReusableCell(withIdentifier: "StartTrainingCell", for: indexPath)
            
        default:
            cell = UITableViewCell()
            break
        }
    
        return cell
    }
    

    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        return indexPath.section == 1
    }
    
    override func tableView(_ tableView: UITableView, canMoveRowAt indexPath: IndexPath) -> Bool {
        return indexPath.section == 1
    }

    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCellEditingStyle, forRowAt indexPath: IndexPath) {
        if editingStyle != .delete {
            return
        }
        
        // Remove exercise object from the training
        self.training.removeExerciseAtIndex(indexPath.row)
        
        // Remove data from TableView
        tableView.deleteRows(at: [indexPath], with: .fade)
        
        CoreData.save()
    }

    override func tableView(_ tableView: UITableView, moveRowAt fromIndexPath: IndexPath, to toIndexPath: IndexPath) {
        // Move object in CoreData
        self.training!.moveExerciseFromIndex(fromIndexPath.row, toIndex: toIndexPath.row)
        CoreData.save()
    }

    override func tableView(_ tableView: UITableView, targetIndexPathForMoveFromRowAt sourceIndexPath: IndexPath, toProposedIndexPath proposedDestinationIndexPath: IndexPath) -> IndexPath {
        if proposedDestinationIndexPath.section != 1 {
            return IndexPath(item: sourceIndexPath.row, section: 1)
        }
        
        return proposedDestinationIndexPath
    }
    
    override func tableView(_ tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
        switch section {
        case 0:
            return "TRAININGTABLEVIEWCONTROLLER_NAME_HEADER".localized
        case 1:
            return "TRAININGTABLEVIEWCONTROLLER_EXERCISES_HEADER".localized
        default:
            return nil
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        // Start training
        if indexPath.section == 2 && indexPath.row == 0 {
            tableView.deselectRow(at: indexPath, animated: true)
            self.startTraining(self.training)
        }
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        if let controller = segue.destination as? ExerciseTableViewController, segue.identifier != nil {
            switch segue.identifier! {
            case "Add":
                controller.exercise = self.training.createExercise()
                
            case "Edit" where sender is UITableViewCell:
                let index = self.tableView.indexPath(for: sender as! UITableViewCell)!.row
                controller.exercise = self.training.exercises!.object(at: index) as! Exercise
                
            default:
                break
            }
        }
    }
    
    func inputTableViewCell(_ cell: InputTableViewCell, didChangedText newText: String) {
        self.training.name = newText
    }
    
    func startTraining(_ training: Training) {
        if self.training.exercises!.count == 0 {
            let controller = UIAlertController(title: "ERROR".localized, message: "TRAININGTABLEVIEWCONTROLLER_NO_EXERCISES".localized, preferredStyle: .alert)
            controller.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
            
            return self.present(controller, animated: true, completion: nil)
        }
        
        let activeTraining = CoreData.shared.createActiveTrainingOrGetActive()
        activeTraining.startTraining(training)
        self.activeTrainingController = ActiveTrainingController(withStoryboard: self.storyboard!, andActiveTraining: activeTraining)
        self.activeTrainingController!.startViaController(self)
        CoreData.save()
    }
}
